From 525d993d345ee1ab2f242c5043875f823b4ebf95 Mon Sep 17 00:00:00 2001
From: Guy Nisan <gnisan@habana.ai>
Date: Wed, 20 Nov 2024 20:05:41 +0200
Subject: [PATCH] [SW-207692] common: gaudi3: restore ptrs on iccm1

In case that arcboot ran with the following cold straps:
1. boot_info_set_done=0
2. validation_pass=1

Then ICCM1 no longer holds the boot ptrs data. This is due to
the fact that bootROM copies boot info manifest into it.
In that case, restore PTRs data by copying it from flash.

Change-Id: I7673b47849996f8f9cdaeb49c09501381e377db0
Signed-off-by: Guy Nisan <gnisan@habana.ai>
---

diff --git a/arch/gaudi3/arc/common.c b/arch/gaudi3/arc/common.c
index 05e295f..0718c37 100644
--- a/arch/gaudi3/arc/common.c
+++ b/arch/gaudi3/arc/common.c
@@ -1504,12 +1504,35 @@
 	return ((sw_status_0_reg & SW_STATUS_SPI_BOOT_DESC_DMA_START_MASK) != 0);
 }
 
+/* In case that arcboot ran with the following cold straps:
+ * 1. boot_info_set_done=0
+ * 2. validation_pass=1
+ *
+ * Then ICCM1 no longser holds the boot ptrs data. This is due to
+ * the fact that we copy boot info manifest into it.
+ * In that case, restore PTRs data by copying it from flash.
+ */
+static void ppboot_ptrs_fixup(void)
+{
+	u32 size;
+	u32 magic;
+
+	memcpy(&magic, (void *)(uintptr_t)IMG_PTRS_MAGIC_OFFSET, sizeof(u32));
+	if (magic == HL_FW_BOOT_MAGIC)
+		return;
+
+	spi_flash_copy(&size, HL_FW_BOOT_PTRS_BASE, sizeof(u32), NULL);
+	ppboot_cp_image(ARC1_ICCM_FULL_BASE, HL_FW_BOOT_PTRS_BASE, size, NULL);
+}
+
 void ppboot_compute_img_ptr_hash(u8 *hash)
 {
 	u32 img_ptrs_size, img_ptrs_img_size, img_ptrs_pub_key_size, img_ptrs_sig_size,
 	img_ptrs_img_offset;
 	u8 crc_bytes, num_crc, img_ptrs_version, img_ptrs_hdr_size;
 
+	ppboot_ptrs_fixup();
+
 	memcpy(&img_ptrs_size, (void *)(uintptr_t)IMG_PTRS_SIZE_OFFSET, 4);
 	memcpy(&img_ptrs_sig_size, (void *)(uintptr_t)IMG_PTRS_SIG_SIZE_OFFSET, 4);
 	memcpy(&img_ptrs_pub_key_size, (void *)(uintptr_t)IMG_PTRS_PUB_KEY_SIZE_OFFSET, 4);
diff --git a/arch/gaudi3/arc/common.h b/arch/gaudi3/arc/common.h
index c7e354d..4d2b9b8 100644
--- a/arch/gaudi3/arc/common.h
+++ b/arch/gaudi3/arc/common.h
@@ -132,6 +132,8 @@
 					offsetof(struct hl_fw_boot_ptrs, version))
 #define IMG_PTRS_HDR_SIZE_OFFSET		(IMG_PTRS_OFFSET + \
 					offsetof(struct hl_fw_boot_ptrs, boot_ptrs_sz))
+#define IMG_PTRS_MAGIC_OFFSET		(IMG_PTRS_OFFSET + \
+					offsetof(struct hl_fw_boot_ptrs, magic))
 
 #define HL_FW_IMG_PTR_VERSION_0		0
 
